fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/DataInKeySpacePath.java

```diff
@@ -37,24 +38,45 @@
 public class DataInKeySpacePath {
 
     @Nonnull
-    private final CompletableFuture<ResolvedKeySpacePath> resolvedPath;
+    private final KeySpacePath path;
+    @Nullable
+    private final Tuple remainder;
     @Nonnull
     private final byte[] value;
 
-    public DataInKeySpacePath(KeySpacePath path, KeyValue rawKeyValue, FDBRecordContext context) {
-        this.resolvedPath = path.toResolvedPathAsync(context, rawKeyValue.getKey());
-        this.value = rawKeyValue.getValue();
-        if (this.value == null) {
+    public DataInKeySpacePath(@Nonnull final KeySpacePath path, @Nullable final Tuple remainder,
+                              @Nullable final byte[] value) {
+        this.path = path;
+        this.remainder = remainder;
+        if (value == null) {
             throw new RecordCoreArgumentException("Value cannot be null")
-                    .addLogInfo(LogMessageKeys.KEY, ByteArrayUtil2.loggable(rawKeyValue.getKey()));
+                    .addLogInfo(LogMessageKeys.KEY, path);
         }
-    }
-
-    public CompletableFuture<ResolvedKeySpacePath> getResolvedPath() {
-        return resolvedPath;
+        this.value = value;
```

Comment:
Instead of copying the `byte[]` here and in `getValue`, I created an issue to convert it to `ByteString`: https://github.com/FoundationDB/fdb-record-layer/issues/3717